Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,632 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos Fernández
+-/
+
+import Mathlib.RingTheory.MvPowerSeries.Evaluation
+import Mathlib.RingTheory.MvPowerSeries.LinearTopology
+import Mathlib.RingTheory.PowerSeries.Basic
+import Mathlib.RingTheory.Nilpotent.Basic
+import Mathlib.Topology.UniformSpace.DiscreteUniformity
+
+/-! # Substitutions in multivariate power series
+
+Here we define the substitution of power series into other power series.
+We follow Bourbaki, Algèbre, chap. 4, §4, n° 3.
+
+Evaluation of a power series at adequate elements has been defined
+in `Mathlib.RingTheory.MvPowerSeries.Evaluation`.
+The goal here is to check the relevant hypotheses:
+* The ring of coefficients is endowed the discrete topology.
+* The main condition rewrites as having vanishing constant coefficient
+* Multivariate power series have a linear topology
+
+-/
+
+namespace MvPowerSeries
+
+variable {σ : Type*}
+  {A : Type*} [CommSemiring A]
+  {R : Type*} [CommRing R] [Algebra A R]
+  {τ : Type*}
+  {S : Type*} [CommRing S] [Algebra A S] [Algebra R S] [IsScalarTower A R S]
+
+open WithPiTopology
+
+/-- Families of power series which can be substituted -/
+structure HasSubst (a : σ → MvPowerSeries τ S) : Prop where
+  const_coeff s : IsNilpotent (constantCoeff τ S (a s))
+  coeff_zero d : {s | (a s).coeff S d ≠ 0}.Finite
+
+variable {a : σ → MvPowerSeries τ S}
+
+lemma coeff_zero_iff :
+    Filter.Tendsto a Filter.cofinite (@nhds _ (@instTopologicalSpace τ S ⊥) 0) ↔
+      ∀ d : τ →₀ ℕ, {s | (a s).coeff S d ≠ 0}.Finite := by
+  letI : UniformSpace S := ⊥
+  simp_rw [tendsto_iff_coeff_tendsto, coeff_zero]
+  apply forall_congr'
+  intro d
+  simp [nhds_discrete]
```

Comment:
```suggestion
local instance foo : UniformSpace S := ⊥

lemma coeff_zero_iff :
    Filter.Tendsto a Filter.cofinite (nhds 0) ↔
      ∀ d : τ →₀ ℕ, {s | (a s).coeff S d ≠ 0}.Finite := by
  simp_rw [tendsto_iff_coeff_tendsto, coeff_zero]
  apply forall_congr'
  intro d
  simp [nhds_discrete]
```
If you add this local instance (with a better name and a comment explaining its point) you can remove all `let : UniformSpace ...` from the file. It has the effect that all rings become topological spaces *in this file* (this should be explained clearly), but not in files importing it, so it seems a good compromise.

Reply:
No, that seem to work.